[NNPA] Change conditions to apply CPU-padding to fix issue #1775.

negiyas opened the pull request:
Change conditions to apply CPU-padding to fix issue #1775.

Signed-off-by: Yasushi Negishi <negishi@jp.ibm.com>

cjvolzka commented:
I tried a build using https://github.com/onnx/onnx-mlir/pull/1777/commits/67fc796fde89c9739c03a32e3243102a13a6f781 
From the original issue, the following models compiled successful:
```
yolov3-10
yolov3-12
candy-9
mosaic-9
pointilism-9
rain-princess-9
udnie-9
```

The following still failed:
```
bvlcalexnet-6
bvlcalexnet-7
bvlcalexnet-8
bvlcalexnet-9
caffenet-6
caffenet-7
caffenet-8
caffenet-9
caffenet-12
efficientnet-lite4-11
mobilenetv2-7
rcnn-ilsvrc13-6
rcnn-ilsvrc13-7
rcnn-ilsvrc13-8
rcnn-ilsvrc13-9
shufflenet-6
shufflenet-7
shufflenet-8
shufflenet-9
shufflenet-v2-10
shufflenet-v2-12

negiyas commented:
@cjvolzka Thank you for testing.
I updated this PR again, and confirmed that this PR works for all models.

negiyas commented:
Test results are attached below.
```
$ cat modelzoo-test.sh

TIME=`date +%Y%m%d-%H%M%S`

ONNX_MLIR=/Dreamer/trl/negishi/src/dlc.git/onnx-mlir
ONNXMLIR=$ONNX_MLIR/build/Debug/bin/onnx-mlir
MODELDIR=models

ONNX_LIST="bvlcalexnet-6 bvlcalexnet-7 bvlcalexnet-8 bvlcalexnet-9 bvlcalexnet-12 caffenet-6 caffenet-7 caffenet-8 caffenet-9 caffenet-12 efficientnet-lite4-11 mobilenet-v2-7 rcnn-ilsvrc13-6 rcnn-ilsvrc13-7 rcnn-ilsvrc13-8 rcnn-ilsvrc13-9 shufflenet-6 shufflenet-7 shufflenet-8 shufflenet-9 shufflenet-v2-10 shufflenet-v2-12 yolov3-10 yolov3-12 candy-9 mosaic-9 pointilism-9 rain-princess-9 udnie-9"

for MODEL in $ONNX_LIST
do

echo $MODEL
$ONNXMLIR --EmitLib --O3 --onnx-op-stats=TXT --mtriple=s390x-ibm-loz --mcpu=z16 --maccel=NNPA $MODELDIR/$MODEL.onnx

done 2>&1 | tee ${0%.*}-${TIME}.log
$ ./modelzoo-test.sh
bvlcalexnet-6
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.MaxPool2D          , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 9
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 9
bvlcalexnet-7
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.MaxPool2D          , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 9
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 9
bvlcalexnet-8
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.MaxPool2D          , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 9
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 9
bvlcalexnet-9
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.MaxPool2D          , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 9
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 9
bvlcalexnet-12
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.MaxPool2D          , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 9
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 9
caffenet-6
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 7
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 7
  caffenet-7
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 7
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 7
caffenet-8
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 7
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 7
caffenet-9
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 7
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 7
caffenet-12
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
   onnx.SqueezeV11         , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Softmax            , 1
  zhigh.Stick              , 7
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 7
efficientnet-lite4-11
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Clip               , 61
   onnx.Constant           , 67
   onnx.Conv               , 30
   onnx.Pad                , 30
   onnx.Squeeze            , 1
   onnx.SqueezeV11         , 1
   onnx.Transpose          , 1
   onnx.UnsqueezeV11       , 1
  zhigh.Add                , 23
  zhigh.AvgPool2D          , 1
  zhigh.Conv2D             , 61
  zhigh.MatMul             , 1
  zhigh.Softmax            , 1
  zhigh.Stick              , 34
  zhigh.StickifiedConstant , 124
  zhigh.Unstick            , 34
mobilenet-v2-7
Unable to open or access models/mobilenet-v2-7.onnx
rcnn-ilsvrc13-6
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Stick              , 6
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 6
rcnn-ilsvrc13-7
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Stick              , 6
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 6
rcnn-ilsvrc13-8
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Stick              , 6
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 6
rcnn-ilsvrc13-9
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Constant           , 10
   onnx.Conv               , 3
   onnx.LRN                , 2
   onnx.MaxPoolSingleOut   , 3
   onnx.Pad                , 3
   onnx.Reshape            , 1
  zhigh.Conv2D             , 2
  zhigh.MatMul             , 3
  zhigh.Relu               , 5
  zhigh.Stick              , 6
  zhigh.StickifiedConstant , 10
  zhigh.Unstick            , 6
shufflenet-6
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.AveragePool        , 3
   onnx.Concat             , 3
   onnx.Constant           , 108
   onnx.Conv               , 48
   onnx.MaxPoolSingleOut   , 1
   onnx.Pad                , 49
   onnx.Reshape            , 33
   onnx.SqueezeV11         , 1
   onnx.Transpose          , 16
   onnx.UnsqueezeV11       , 1
  zhigh.Add                , 13
  zhigh.AvgPool2D          , 1
  zhigh.Conv2D             , 1
  zhigh.MatMul             , 1
  zhigh.Relu               , 32
  zhigh.Softmax            , 1
  zhigh.Stick              , 36
  zhigh.StickifiedConstant , 4
  zhigh.Unstick            , 36
shufflenet-7
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.AveragePool        , 3
   onnx.Concat             , 3
   onnx.Constant           , 108
   onnx.Conv               , 48
   onnx.MaxPoolSingleOut   , 1
   onnx.Pad                , 49
   onnx.Reshape            , 33
   onnx.SqueezeV11         , 1
   onnx.Transpose          , 16
   onnx.UnsqueezeV11       , 1
  zhigh.Add                , 13
  zhigh.AvgPool2D          , 1
  zhigh.Conv2D             , 1
  zhigh.MatMul             , 1
  zhigh.Relu               , 32
  zhigh.Softmax            , 1
  zhigh.Stick              , 36
  zhigh.StickifiedConstant , 4
  zhigh.Unstick            , 36
shufflenet-8
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.AveragePool        , 3
   onnx.Concat             , 3
   onnx.Constant           , 108
   onnx.Conv               , 48
   onnx.MaxPoolSingleOut   , 1
   onnx.Pad                , 49
   onnx.Reshape            , 33
   onnx.SqueezeV11         , 1
   onnx.Transpose          , 16
   onnx.UnsqueezeV11       , 1
  zhigh.Add                , 13
  zhigh.AvgPool2D          , 1
  zhigh.Conv2D             , 1
  zhigh.MatMul             , 1
  zhigh.Relu               , 32
  zhigh.Softmax            , 1
  zhigh.Stick              , 36
  zhigh.StickifiedConstant , 4
  zhigh.Unstick            , 36
shufflenet-9
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.AveragePool        , 3
   onnx.Concat             , 3
   onnx.Constant           , 108
   onnx.Conv               , 48
   onnx.MaxPoolSingleOut   , 1
   onnx.Pad                , 49
   onnx.Reshape            , 33
   onnx.SqueezeV11         , 1
   onnx.Transpose          , 16
   onnx.UnsqueezeV11       , 1
  zhigh.Add                , 13
  zhigh.AvgPool2D          , 1
  zhigh.Conv2D             , 1
  zhigh.MatMul             , 1
  zhigh.Relu               , 32
  zhigh.Softmax            , 1
  zhigh.Stick              , 36
  zhigh.StickifiedConstant , 4
  zhigh.Unstick            , 36
shufflenet-v2-10
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Concat             , 16
   onnx.Constant           , 75
   onnx.Conv               , 19
   onnx.MaxPoolSingleOut   , 1
   onnx.Pad                , 20
   onnx.ReduceMean         , 1
   onnx.Reshape            , 32
   onnx.Split              , 13
   onnx.Transpose          , 16
  zhigh.Conv2D             , 37
  zhigh.MatMul             , 1
  zhigh.Stick              , 38
  zhigh.StickifiedConstant , 76
  zhigh.Unstick            , 38
shufflenet-v2-12
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Concat             , 16
   onnx.Constant           , 75
   onnx.Conv               , 19
   onnx.MaxPoolSingleOut   , 1
   onnx.Pad                , 20
   onnx.ReduceMean         , 1
   onnx.Reshape            , 32
   onnx.Split              , 13
   onnx.Transpose          , 16
  zhigh.Conv2D             , 37
  zhigh.MatMul             , 1
  zhigh.Stick              , 38
  zhigh.StickifiedConstant , 76
  zhigh.Unstick            , 38
yolov3-10
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Add                , 41
   onnx.Cast               , 30
   onnx.Ceil               , 7
   onnx.Concat             , 24
   onnx.Constant           , 179
   onnx.Conv               , 75
   onnx.Dim                , 6
   onnx.Div                , 15
   onnx.Exp                , 3
   onnx.LeakyRelu          , 72
   onnx.Loop               , 6
   onnx.Mul                , 17
   onnx.NonMaxSuppression  , 1
   onnx.ReduceMin          , 1
   onnx.Reshape            , 15
   onnx.Resize             , 2
   onnx.Return             , 6
   onnx.Sigmoid            , 9
   onnx.Slice              , 31
   onnx.Squeeze            , 7
   onnx.Sub                , 6
   onnx.Tile               , 6
   onnx.Transpose          , 4
   onnx.Unsqueeze          , 2
  zhigh.Add                , 1
  zhigh.Div                , 4
  zhigh.Stick              , 4
  zhigh.StickifiedConstant , 2
  zhigh.Sub                , 1
  zhigh.Unstick            , 4
yolov3-12
Operations encountered:
-----------------------
   func.func               , 1
   func.return             , 1
   onnx.Add                , 41
   onnx.Cast               , 30
   onnx.Ceil               , 7
   onnx.Concat             , 24
   onnx.Constant           , 178
   onnx.Conv               , 75
   onnx.Dim                , 6
   onnx.Div                , 15
   onnx.Exp                , 3
   onnx.LeakyRelu          , 72
   onnx.Loop               , 6
   onnx.Mul                , 17
   onnx.NoValue            , 1
   onnx.NonMaxSuppression  , 1
   onnx.ReduceMin          , 1
   onnx.Reshape            , 15
   onnx.Resize             , 2
   onnx.Return             , 6
   onnx.Sigmoid            , 9
   onnx.Slice              , 31
   onnx.Squeeze            , 7
   onnx.Sub                , 6
   onnx.Tile               , 6
   onnx.Transpose          , 4
   onnx.Unsqueeze          , 2
  zhigh.Add                , 1
  zhigh.Div                , 4
  zhigh.Stick              , 4
  zhigh.StickifiedConstant , 2
  zhigh.Sub                , 1
  zhigh.Unstick            , 4
candy-9
Operations encountered:
-----------------------
   func.func                  , 1
   func.return                , 1
   onnx.Cast                  , 4
   onnx.Concat                , 5
   onnx.Constant              , 43
   onnx.Conv                  , 3
   onnx.Dim                   , 2
   onnx.Floor                 , 3
   onnx.InstanceNormalization , 15
   onnx.Mul                   , 2
   onnx.NoValue               , 1
   onnx.Pad                   , 16
   onnx.Resize                , 2
   onnx.Squeeze               , 2
   onnx.Unsqueeze             , 3
  zhigh.Add                   , 5
  zhigh.Conv2D                , 13
  zhigh.Div                   , 2
  zhigh.Relu                  , 10
  zhigh.Stick                 , 31
  zhigh.StickifiedConstant    , 27
  zhigh.Unstick               , 30
mosaic-9
Operations encountered:
-----------------------
   func.func                  , 1
   func.return                , 1
   onnx.Cast                  , 4
   onnx.Concat                , 5
   onnx.Constant              , 43
   onnx.Conv                  , 3
   onnx.Dim                   , 2
   onnx.Floor                 , 3
   onnx.InstanceNormalization , 15
   onnx.Mul                   , 2
   onnx.NoValue               , 1
   onnx.Pad                   , 16
   onnx.Resize                , 2
   onnx.Squeeze               , 2
   onnx.Unsqueeze             , 3
  zhigh.Add                   , 5
  zhigh.Conv2D                , 13
  zhigh.Div                   , 2
  zhigh.Relu                  , 10
  zhigh.Stick                 , 31
  zhigh.StickifiedConstant    , 27
  zhigh.Unstick               , 30
pointilism-9
Operations encountered:
-----------------------
   func.func                  , 1
   func.return                , 1
   onnx.Cast                  , 4
   onnx.Concat                , 5
   onnx.Constant              , 43
   onnx.Conv                  , 3
   onnx.Dim                   , 2
   onnx.Floor                 , 3
   onnx.InstanceNormalization , 15
   onnx.Mul                   , 2
   onnx.NoValue               , 1
   onnx.Pad                   , 16
   onnx.Resize                , 2
   onnx.Squeeze               , 2
   onnx.Unsqueeze             , 3
  zhigh.Add                   , 5
  zhigh.Conv2D                , 13
  zhigh.Div                   , 2
  zhigh.Relu                  , 10
  zhigh.Stick                 , 31
  zhigh.StickifiedConstant    , 27
  zhigh.Unstick               , 30
rain-princess-9
Operations encountered:
-----------------------
   func.func                  , 1
   func.return                , 1
   onnx.Cast                  , 4
   onnx.Concat                , 5
   onnx.Constant              , 43
   onnx.Conv                  , 3
   onnx.Dim                   , 2
   onnx.Floor                 , 3
   onnx.InstanceNormalization , 15
   onnx.Mul                   , 2
   onnx.NoValue               , 1
   onnx.Pad                   , 16
   onnx.Resize                , 2
   onnx.Squeeze               , 2
   onnx.Unsqueeze             , 3
  zhigh.Add                   , 5
  zhigh.Conv2D                , 13
  zhigh.Div                   , 2
  zhigh.Relu                  , 10
  zhigh.Stick                 , 31
  zhigh.StickifiedConstant    , 27
  zhigh.Unstick               , 30
udnie-9
Operations encountered:
-----------------------
   func.func                  , 1
   func.return                , 1
   onnx.Cast                  , 4
   onnx.Concat                , 5
   onnx.Constant              , 43
   onnx.Conv                  , 3
   onnx.Dim                   , 2
   onnx.Floor                 , 3
   onnx.InstanceNormalization , 15
   onnx.Mul                   , 2
   onnx.NoValue               , 1
   onnx.Pad                   , 16
   onnx.Resize                , 2
   onnx.Squeeze               , 2
   onnx.Unsqueeze             , 3
  zhigh.Add                   , 5
  zhigh.Conv2D                , 13
  zhigh.Div                   , 2
  zhigh.Relu                  , 10
  zhigh.Stick                 , 31
  zhigh.StickifiedConstant    , 27
  zhigh.Unstick               , 30
```

negiyas commented:
@jenkins-droid test this please

messerb5467 commented:
We missed mobilenetv2-7 and I imagine that's more related to our test systems. I'll poke around and finish out that model.

messerb5467 commented:
@negiyas There was a minor spelling error above in `mobilenetv2-7` having a `-` before main part of the model and the `v2`. This ends up being such a subtle part of the model naming and I wanted to mention something in case someone decides to reuse the one-off script you provided above.

tungld commented on src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:
```diff
@@ -224,9 +224,13 @@ bool canInferencePadsForNNPAConv(ONNXConvOp op) {
+         (llvm::all_of(shapeHelper.pads,
+             [](IndexExpr val) { return val.isLiteral(); })) &&
+         (op.auto_pad() != "VALID") && (inputShape.size() == 4) &&
+         (inputShape[1] > 0) && (inputShape[2] > 0) && (inputShape[3] > 0);
```
Could you split this complex condition into small ones and add a comment for each, like
```C
// comment
if ()
  return false;
// comment
if ()
  return false;
// Otherwise, the op is legal.
return true;
```
It makes the code easy to read and debug. Thanks!

negiyas replied:
Fixed. Thank you.

tungld commented on src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:
```diff
@@ -224,9 +224,13 @@ bool canInferencePadsForNNPAConv(ONNXConvOp op) {
-             shapeHelper.pads, [](IndexExpr val) { return val.isLiteral(); }));
+         (llvm::all_of(shapeHelper.pads,
+             [](IndexExpr val) { return val.isLiteral(); })) &&
+         (op.auto_pad() != "VALID") && (inputShape.size() == 4) &&
```
Should it be `VALID_PADDING`? Better to use `op.auto_pad(). equals_insensitive("VALID_PADDING")` instead of `!=`

negiyas replied:
I updated the code to use equals_insensitive instead of !=. Thank you for confirmation.

In the onnx dialect, auto_pad must be either NOTSET, SAME_UPPER, SAME_LOWER or VALID (not VALID_PADDING). (c.f. https://github.com/onnx/onnx/blob/main/docs/Operators.md#Conv)

tungld commented on test/mlir/accelerators/nnpa/conversion/rewrite-onnx-for-zhigh.mlir:
```diff
@@ -527,3 +527,12 @@ func.func @test_onnx_conv2d_not_insert_onnxpad_when_not_necessary(%arg0: tensor<
+    return %0 : tensor<1x64x?x?xf32>
+  // CHECK-LABEL: test_onnx_conv2d_not_insert_onnxpad_if_cannot_get_pads
+  // CHECK-NOT: "onnx.Pad"
+}
```
Could you add a lit test for the condition related to `VALID_PADDING`?

negiyas replied:
Appended a lit test for the onnx.Conv with auto_pad==VALID case. Thank you.

negiyas commented:
@messerb5467 

> @negiyas There was a minor spelling error above in `mobilenetv2-7` having a `-` before main part of the model and the `v2`. This ends up being such a subtle part of the model naming and I wanted to mention something in case someone decides to reuse the one-off script you provided above.

I confirmed that this PR works for mobilenetv2-7, and fixed the script and test results in my above comment to avoid reuse of the wrong script.
Thank you.

messerb5467 commented:
> @messerb5467
> 
> > @negiyas There was a minor spelling error above in `mobilenetv2-7` having a `-` before main part of the model and the `v2`. This ends up being such a subtle part of the model naming and I wanted to mention something in case someone decides to reuse the one-off script you provided above.
> 
> I confirmed that this PR works for mobilenetv2-7, and fixed the script and test results in my above comment to avoid reuse of the wrong script. Thank you.

You're welcome. Thanks for the hard work. I reviewed the output above and I'm fine with this going through once the approval comes from another maintainer of onnx-mlir.

tungld reviewed (approved):
LGTM. Thanks for the quick fix!

AlexandreEichenberger commented on src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:
```diff
@@ -224,9 +224,22 @@ bool canInferencePadsForNNPAConv(ONNXConvOp op) {
+  if (shapeHelper.pads.size() != 4)
+    return false;
+  // all dimensions of pad should be literal
+  if (llvm::any_of(
```
@negiyas Are these constrains (no valid, image size being compile time constant) needed because there is a bug otherwise, or it is a real constraints, namely we cannot use CPU padding in such case?

Naively, I would have assumed that we can always do CPU padding.

This PR is fine, I would like to understand if long term we should have a better fix.
